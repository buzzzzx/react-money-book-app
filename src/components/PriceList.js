import React from "react";

const PriceList = ({ items, onModifyItem, onDeleteItem }) => (
  <ul className="list-group list-group-flush">
    {items.map((item) => (
      <li
        className="list-group-item d-flex justify-content-between align-items-center"
        key={item.id}
      >
        <span className="col-1 badge badge-primary">{item.category.name}</span>
        <span className="col-2">{item.title}</span>
        <span className="col-5 font-weight-bold">
          {item.category.type === "income" ? "+" : "-"}
          {item.price}
        </span>
        <span className="col-2">{item.date}</span>
        <button className="btn btn-primary" onClick={() => onModifyItem(item)}>
          编辑
        </button>
        <button className="btn btn-danger" onClick={() => onDeleteItem(item)}>
          删除
        </button>
      </li>
    ))}
  </ul>
);

export default PriceList;
