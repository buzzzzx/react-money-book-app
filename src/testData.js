export const testCategories = [
  {
    name: "旅行",
    iconName: "ios-plane",
    id: "1",
    type: "outcome",
  },
  {
    name: "餐饮",
    iconName: "ios-restaurant",
    id: "2",
    type: "outcome",
  },
  {
    name: "购物",
    iconName: "ios-basket",
    id: "3",
    type: "outcome",
  },
  {
    name: "数码",
    iconName: "ios-phone-portrait",
    id: "4",
    type: "outcome",
  },
  {
    name: "工资",
    iconName: "ios-card",
    id: "10",
    type: "income",
  },
  {
    name: "兼职",
    iconName: "ios-cash",
    id: "11",
    type: "income",
  },
  {
    name: "理财",
    iconName: "logo-yen",
    id: "12",
    type: "income",
  },
];

export const testItems = [
  {
    title: "buy stuff for kitten",
    price: 100,
    date: "2020-07-15",
    monthCategory: "2020-7",
    id: "_kly1klf4g",
    cid: "1",
    timestamp: 1534291200000,
  },
  {
    title: "这是我的工资",
    price: 20000,
    date: "2020-08-18",
    monthCategory: "2020-8",
    id: "_bd16bjeen",
    cid: "2",
    timestamp: 1534550400000,
  },
  {
    title: "和哥们一起喝酒",
    price: 300,
    date: "2020-08-20",
    monthCategory: "2020-8",
    id: "_jjfice21k",
    cid: "3",
    timestamp: 1534723200000,
  },
  {
    title: "理财收入",
    price: 1000,
    date: "2020-08-11",
    monthCategory: "2020-8",
    id: "_1fg1wme63",
    cid: "11",
    timestamp: 1533945600000,
  },
  {
    title: "理财收入",
    price: 300,
    date: "2020-11-15",
    monthCategory: "2020-11",
    id: "_qryggm5y8",
    cid: "12",
    timestamp: 1534291200000,
  },
  {
    title: "请别人吃饭",
    price: 300,
    date: "2020-11-15",
    monthCategory: "2020-11",
    id: "_qryggm511",
    cid: "3",
    timestamp: 1534291200000,
  },
];
