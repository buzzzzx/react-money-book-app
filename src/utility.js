export const LIST_VIEW = "list";
export const CHART_VIEW = "chart";
