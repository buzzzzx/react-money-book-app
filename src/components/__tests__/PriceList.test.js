import React from "react";
import { shallow } from "enzyme";
import Ionicon from "react-ionicons";
import { Link } from "@reach/router";

import PriceList from "../PriceList";
import { flattenData } from "../../utility";
import { testCategories, testItems } from "../../testData";

const categories = flattenData(testCategories);
const itemsWithCategory = testItems.map((item) => {
  item.category = categories[item.cid];
  return item;
});

const props = {
  items: itemsWithCategory,
  onDeleteItem: jest.fn(),
};

let wrapper;
describe("test <PriceList />", () => {
  // this will be called before every test case(it)
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props} />);
  });

  it("should render component to match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render price items to correct length", () => {
    expect(wrapper.find(".list-group-item").length).toEqual(props.items.length);
  });

  it("should render correct icon and content of price items", () => {
    const iconList = wrapper.find(".list-group-item").first().find(Ionicon);
    // number of icons for each price item
    expect(iconList.length).toEqual(3);
    // category icon name
    expect(iconList.first().props().icon).toEqual(
      props.items[0].category.iconName
    );
  });

  it("should trigger the correct callback", () => {
    const firstItem = wrapper.find(".list-group-item").first();
    // trigger modify event on first price item
    firstItem.find(Link).simulate("click");
    // expect(props.onModifyItem).toHaveBeenCalledWith(props.items[0]);
    // trigger delete event on first price item
    firstItem.find("a").first().simulate("click");
    expect(props.onDeleteItem).toHaveBeenCalledWith(props.items[0]);
  });
});
