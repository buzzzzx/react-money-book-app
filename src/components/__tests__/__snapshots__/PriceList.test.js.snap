// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <PriceList /> should render component to match snapshot 1`] = `ShallowWrapper {}`;
